File: warehouse_restriction/events/material_request.py
import frappe

@frappe.whitelist()
def set_warehouse(doctype, txt, searchfield, start, page_len, filters):
    warehouse_list = []
    
    company = filters.get('company') if filters else None
    
    if frappe.db.exists("Warehouse Restriction", {'user': frappe.session.user}):
        r_warehouse_list = frappe.get_list("Warehouse Restriction", {'user': frappe.session.user, 'document_name': "Material Request"}, ['name'])
        warehouse = frappe.get_doc("Warehouse Restriction", r_warehouse_list[0].name)
        
        unique_warehouse = {}
        
        for row in warehouse.warehouses:
            if row.warehouse not in unique_warehouse:
                unique_warehouse[row.warehouse] = row.warehouse
                warehouse_list.append((row.warehouse, frappe.get_value("Warehouse", row.warehouse, "warehouse_name")))
    else:
        if company:
            warehouses = frappe.get_all("Warehouse", filters={'company': company}, fields=['name', 'warehouse_name'])
            for warehouse in warehouses:
                warehouse_list.append((warehouse.name, warehouse.warehouse_name))
    
    return warehouse_list

def restrictwarehouse(doc,method=None):
    warehouse=doc.set_warehouse
    if not warehouse:
        warehouse = doc.items[0].warehouse
    if frappe.db.exists("Warehouse Restriction", {'user': frappe.session.user,'document_name': "Material Request"}):
        
        restricted_warehouse = frappe.get_doc("Warehouse Restriction", frappe.session.user+"-Material Request") 
        exist = 0
        for row in restricted_warehouse.warehouses:
            if row.warehouse == warehouse:
                exist += 1
        if not exist:
            frappe.throw(f"Warehouse {warehouse} is not asigned current user. ")  
    
    return 0

File: warehouse_restriction/public/js/material_request.js
frappe.ui.form.on("Material Request", {
    onload: function(frm) {
        frm.set_query("set_warehouse", () => {
            return {
                query: "warehouse_restriction.events.material_request.set_warehouse",
                filters: {
                    company: frm.doc.company
                }
            };
        });

        frm.set_query('warehouse', 'items', () => {
            return {
                query: "warehouse_restriction.events.material_request.set_warehouse",
                filters: {
                    company: frm.doc.company
                }
            };
        });
    },

    set_warehouse: function(frm) {
        if (frm.doc.set_warehouse) {
            console.log(frm.doc.set_warehouse, frappe.session.user+"-Material Request");
            
            frappe.db.exists('Warehouse Restriction',frappe.session.user+"-Material Request")
            .then(exists => {
                console.log(exists);
                
                if(exists){
                    frappe.db.get_doc('Warehouse Restriction', null, { 'user': frappe.session.user,'document_name':"Material Request"})
                    .then(doc => {    
                        let w_exist = 0
                        for(let i of doc.warehouses){
                            if(i.warehouse == frm.doc.set_warehouse){
                                w_exist += 1
                            }
                            if( w_exist == 0 ){
                                frappe.msgprint(__('You are not assignend to the selected warehouse {0}.', [frm.doc.set_warehouse]));
                                frappe.model.set_value("Material Request",frm.doc.name,"set_warehouse","")
                            }
                        }
                    })
                }
            })
        }
    },
});
